File: bridge/pkg/sidecar/token.go
// Package sidecar provides token generation for secure sidecar communication
package sidecar

import (
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"fmt"
	"strconv"
	"strings"
	"time"
)

const (
	// TokenTTL is the token time-to-live (30 minutes per Oracle recommendation)
	TokenTTL = 30 * time.Minute

	// MaxTimestampAge is the maximum age of a token timestamp (5 minutes)
	// Tokens with timestamps older than this will be rejected
	MaxTimestampAge = 5 * time.Minute
)

// TokenGenerator generates ephemeral tokens for sidecar authentication
type TokenGenerator struct {
	sharedSecret []byte
}

// NewTokenGenerator creates a new TokenGenerator with the given shared secret
func NewTokenGenerator(sharedSecret []byte) (*TokenGenerator, error) {
	if len(sharedSecret) == 0 {
		return nil, fmt.Errorf("shared secret cannot be empty")
	}
	return &TokenGenerator{
		sharedSecret: sharedSecret,
	}, nil
}

// GenerateToken generates an ephemeral token for the given request
// Token format: {request_id}:{timestamp}:{operation}:{signature}
// The signature is HMAC-SHA256 of (request_id + timestamp + operation)
func (tg *TokenGenerator) GenerateToken(requestID string, operation string) (string, int64, error) {
	if requestID == "" {
		return "", 0, fmt.Errorf("request_id cannot be empty")
	}
	if operation == "" {
		return "", 0, fmt.Errorf("operation cannot be empty")
	}

	timestamp := time.Now().Unix()

	// Create the data to sign: request_id + timestamp + operation
	dataToSign := fmt.Sprintf("%s%d%s", requestID, timestamp, operation)

	// Calculate HMAC-SHA256 signature
	signature := tg.calculateHMAC(dataToSign)

	// Format token as: request_id:timestamp:operation:signature
	token := fmt.Sprintf("%s:%d:%s:%s", requestID, timestamp, operation, signature)

	return token, timestamp, nil
}

// calculateHMAC calculates the HMAC-SHA256 of the given data
func (tg *TokenGenerator) calculateHMAC(data string) string {
	h := hmac.New(sha256.New, tg.sharedSecret)
	h.Write([]byte(data))
	return hex.EncodeToString(h.Sum(nil))
}

// TokenInfo represents the parsed components of a token
type TokenInfo struct {
	RequestID  string
	Timestamp  int64
	Operation  string
	Signature  string
	Expiration int64 // Unix timestamp when token expires
}

// ParseToken parses a token into its components
func ParseToken(token string) (*TokenInfo, error) {
	parts := strings.Split(token, ":")
	if len(parts) != 4 {
		return nil, fmt.Errorf("invalid token format: expected 4 parts, got %d", len(parts))
	}

	timestamp, err := strconv.ParseInt(parts[1], 10, 64)
	if err != nil {
		return nil, fmt.Errorf("invalid timestamp: %w", err)
	}

	return &TokenInfo{
		RequestID:  parts[0],
		Timestamp:  timestamp,
		Operation:  parts[2],
		Signature:  parts[3],
		Expiration: timestamp + int64(TokenTTL.Seconds()),
	}, nil
}

// GenerateRequestID generates a unique request ID using crypto/rand
func GenerateRequestID() (string, error) {
	b := make([]byte, 16)
	if _, err := rand.Read(b); err != nil {
		return "", fmt.Errorf("failed to generate request ID: %w", err)
	}
	return hex.EncodeToString(b), nil
}

// ValidateTokenSignature validates the signature of a token
func (tg *TokenGenerator) ValidateTokenSignature(token string) (bool, error) {
	info, err := ParseToken(token)
	if err != nil {
		return false, err
	}

	// Recreate the data that was signed
	dataToSign := fmt.Sprintf("%s%d%s", info.RequestID, info.Timestamp, info.Operation)

	// Calculate the expected signature
	expectedSignature := tg.calculateHMAC(dataToSign)

	// Use constant-time comparison to prevent timing attacks
	return hmac.Equal([]byte(info.Signature), []byte(expectedSignature)), nil
}

// IsTokenExpired checks if a token has expired based on its TTL
func IsTokenExpired(info *TokenInfo) bool {
	now := time.Now().Unix()
	return now > info.Expiration
}

// IsTokenTooOld checks if a token's timestamp is too old (beyond MaxTimestampAge)
func IsTokenTooOld(info *TokenInfo) bool {
	now := time.Now().Unix()
	maxAge := int64(MaxTimestampAge.Seconds())
	return (now - info.Timestamp) > maxAge
}

// ValidateToken performs full validation of a token including signature, expiration, and age
func (tg *TokenGenerator) ValidateToken(token string) (*TokenInfo, error) {
	info, err := ParseToken(token)
	if err != nil {
		return nil, err
	}

	// Check if token is too old (timestamp > 5 minutes ago)
	if IsTokenTooOld(info) {
		return nil, fmt.Errorf("token timestamp is too old (> %s)", MaxTimestampAge)
	}

	// Check if token has expired (TTL exceeded)
	if IsTokenExpired(info) {
		return nil, fmt.Errorf("token has expired (TTL: %s)", TokenTTL)
	}

	// Verify signature
	valid, err := tg.ValidateTokenSignature(token)
	if err != nil {
		return nil, fmt.Errorf("signature validation failed: %w", err)
	}
	if !valid {
		return nil, fmt.Errorf("invalid token signature")
	}

	return info, nil
}

// GenerateSharedSecret generates a random shared secret for development/testing
// In production, this should be loaded from the keystore
func GenerateSharedSecret() ([]byte, error) {
	secret := make([]byte, 32) // 256 bits
	if _, err := rand.Read(secret); err != nil {
		return nil, fmt.Errorf("failed to generate shared secret: %w", err)
	}
	return secret, nil
}

// EncodeSharedSecret encodes a shared secret to base64 for storage/transmission
func EncodeSharedSecret(secret []byte) string {
	return base64.StdEncoding.EncodeToString(secret)
}

// DecodeSharedSecret decodes a base64-encoded shared secret
func DecodeSharedSecret(encoded string) ([]byte, error) {
	return base64.StdEncoding.DecodeString(encoded)
}

File: sidecar/src/security/token.rs
use hmac::{Hmac, Mac};
use sha2::Sha256;
use std::time::{Duration, SystemTime, UNIX_EPOCH};

type HmacSha256 = Hmac<Sha256>;

pub const TOKEN_TTL_SECONDS: u64 = 1800; // 30 minutes
pub const MAX_TIMESTAMP_AGE_SECONDS: u64 = 300; // 5 minutes

#[derive(Debug, Clone, PartialEq)]
pub struct TokenInfo {
    pub request_id: String,
    pub timestamp: i64,
    pub operation: String,
    pub signature: String,
    pub expiration: i64,
}

#[derive(Debug, thiserror::Error)]
pub enum TokenError {
    #[error("invalid token format: expected 4 parts, got {0}")]
    InvalidFormat(usize),
    #[error("invalid timestamp: {0}")]
    InvalidTimestamp(#[from] std::num::ParseIntError),
    #[error("token timestamp is too old (> {0:?})")]
    TokenTooOld(Duration),
    #[error("token has expired (TTL: {0:?})")]
    TokenExpired(Duration),
    #[error("invalid token signature")]
    InvalidSignature,
    #[error("HMAC error: {0}")]
    HmacError(String),
}

/// Parses a token into its components.
/// Token format: {request_id}:{timestamp}:{operation}:{signature}
pub fn parse_token(token: &str) -> Result<TokenInfo, TokenError> {
    let parts: Vec<&str> = token.split(':').collect();

    if parts.len() != 4 {
        return Err(TokenError::InvalidFormat(parts.len()));
    }

    let timestamp = parts[1].parse::<i64>()?;
    let expiration = timestamp + TOKEN_TTL_SECONDS as i64;

    Ok(TokenInfo {
        request_id: parts[0].to_string(),
        timestamp,
        operation: parts[2].to_string(),
        signature: parts[3].to_string(),
        expiration,
    })
}

/// Validates a token's signature using HMAC-SHA256.
pub fn validate_token_signature(
    token: &str,
    shared_secret: &[u8],
) -> Result<bool, TokenError> {
    let info = parse_token(token)?;

    let data_to_sign = format!("{}{}{}", info.request_id, info.timestamp, info.operation);

    let mut mac = HmacSha256::new_from_slice(shared_secret)
        .map_err(|e| TokenError::HmacError(e.to_string()))?;

    mac.update(data_to_sign.as_bytes());

    let expected_signature = hex::encode(mac.finalize().into_bytes());

    Ok(info.signature == expected_signature)
}

/// Checks if a token has expired based on its TTL.
pub fn is_token_expired(info: &TokenInfo) -> bool {
    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_secs() as i64;

    now > info.expiration
}

/// Checks if a token's timestamp is too old (beyond MAX_TIMESTAMP_AGE_SECONDS).
pub fn is_token_too_old(info: &TokenInfo) -> bool {
    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_secs() as i64;

    let age = now - info.timestamp;
    age > MAX_TIMESTAMP_AGE_SECONDS as i64
}

/// Performs full validation of a token including signature, expiration, and age.
pub fn validate_token(
    token: &str,
    shared_secret: &[u8],
) -> Result<TokenInfo, TokenError> {
    let info = parse_token(token)?;

    // Check if token is too old (timestamp > 5 minutes ago)
    if is_token_too_old(&info) {
        return Err(TokenError::TokenTooOld(Duration::from_secs(
            MAX_TIMESTAMP_AGE_SECONDS,
        )));
    }

    // Check if token has expired (TTL exceeded)
    if is_token_expired(&info) {
        return Err(TokenError::TokenExpired(Duration::from_secs(TOKEN_TTL_SECONDS)));
    }

    // Verify signature
    let valid = validate_token_signature(token, shared_secret)?;
    if !valid {
        return Err(TokenError::InvalidSignature);
    }

    Ok(info)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_token_valid() {
        let token = "test-request-id:1234567890:test-operation:abcdef123456";
        let result = parse_token(token);

        assert!(result.is_ok());
        let info = result.unwrap();
        assert_eq!(info.request_id, "test-request-id");
        assert_eq!(info.timestamp, 1234567890);
        assert_eq!(info.operation, "test-operation");
        assert_eq!(info.signature, "abcdef123456");
    }

    #[test]
    fn test_parse_token_invalid_format() {
        let token = "invalid:token";
        let result = parse_token(token);

        assert!(result.is_err());
        match result {
            Err(TokenError::InvalidFormat(n)) => assert_eq!(n, 2),
            _ => panic!("Expected InvalidFormat error"),
        }
    }

    #[test]
    fn test_parse_token_invalid_timestamp() {
        let token = "test-request-id:not-a-number:test-operation:abcdef";
        let result = parse_token(token);

        assert!(result.is_err());
        assert!(matches!(result, Err(TokenError::InvalidTimestamp(_))));
    }

    #[test]
    fn test_validate_token_signature_valid() {
        let shared_secret = b"test-secret-key-32-bytes-long!";
        let request_id = "test-request-id";
        let timestamp = 1234567890;
        let operation = "test-operation";

        let data_to_sign = format!("{}{}{}", request_id, timestamp, operation);

        let mut mac = HmacSha256::new_from_slice(shared_secret).unwrap();
        mac.update(data_to_sign.as_bytes());
        let signature = hex::encode(mac.finalize().into_bytes());

        let token = format!("{}:{}:{}:{}", request_id, timestamp, operation, signature);

        let result = validate_token_signature(&token, shared_secret);
        assert!(result.is_ok());
        assert!(result.unwrap());
    }

    #[test]
    fn test_validate_token_signature_invalid() {
        let shared_secret = b"test-secret-key-32-bytes-long!";
        let token = "test-request-id:1234567890:test-operation:invalid-signature";

        let result = validate_token_signature(token, shared_secret);
        assert!(result.is_ok());
        assert!(!result.unwrap());
    }

    #[test]
    fn test_is_token_expired_true() {
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_secs() as i64;

        let info = TokenInfo {
            request_id: "test".to_string(),
            timestamp: now - TOKEN_TTL_SECONDS as i64 - 100,
            operation: "test".to_string(),
            signature: "test".to_string(),
            expiration: now - 100,
        };

        assert!(is_token_expired(&info));
    }

    #[test]
    fn test_is_token_expired_false() {
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_secs() as i64;

        let info = TokenInfo {
            request_id: "test".to_string(),
            timestamp: now,
            operation: "test".to_string(),
            signature: "test".to_string(),
            expiration: now + TOKEN_TTL_SECONDS as i64,
        };

        assert!(!is_token_expired(&info));
    }

    #[test]
    fn test_is_token_too_old_true() {
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_secs() as i64;

        let info = TokenInfo {
            request_id: "test".to_string(),
            timestamp: now - MAX_TIMESTAMP_AGE_SECONDS as i64 - 100,
            operation: "test".to_string(),
            signature: "test".to_string(),
            expiration: now + 1000,
        };

        assert!(is_token_too_old(&info));
    }

    #[test]
    fn test_is_token_too_old_false() {
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_secs() as i64;

        let info = TokenInfo {
            request_id: "test".to_string(),
            timestamp: now,
            operation: "test".to_string(),
            signature: "test".to_string(),
            expiration: now + 1000,
        };

        assert!(!is_token_too_old(&info));
    }

    #[test]
    fn test_validate_token_success() {
        let shared_secret = b"test-secret-key-32-bytes-long!";
        let request_id = "test-request-id";
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_secs() as i64;
        let timestamp = now;
        let operation = "test-operation";

        let data_to_sign = format!("{}{}{}", request_id, timestamp, operation);

        let mut mac = HmacSha256::new_from_slice(shared_secret).unwrap();
        mac.update(data_to_sign.as_bytes());
        let signature = hex::encode(mac.finalize().into_bytes());

        let token = format!("{}:{}:{}:{}", request_id, timestamp, operation, signature);

        let result = validate_token(&token, shared_secret);
        assert!(result.is_ok());
    }

    #[test]
    fn test_validate_token_too_old() {
        let shared_secret = b"test-secret-key-32-bytes-long!";
        let request_id = "test-request-id";
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_secs() as i64;
        let timestamp = now - MAX_TIMESTAMP_AGE_SECONDS as i64 - 100;
        let operation = "test-operation";

        let data_to_sign = format!("{}{}{}", request_id, timestamp, operation);

        let mut mac = HmacSha256::new_from_slice(shared_secret).unwrap();
        mac.update(data_to_sign.as_bytes());
        let signature = hex::encode(mac.finalize().into_bytes());

        let token = format!("{}:{}:{}:{}", request_id, timestamp, operation, signature);

        let result = validate_token(&token, shared_secret);
        assert!(result.is_err());
        assert!(matches!(result, Err(TokenError::TokenTooOld(_))));
    }

    #[test]
    fn test_validate_token_expired() {
        let shared_secret = b"test-secret-key-32-bytes-long!";
        let request_id = "test-request-id";
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_secs() as i64;
        let timestamp = now - TOKEN_TTL_SECONDS as i64 - 100;
        let operation = "test-operation";

        let data_to_sign = format!("{}{}{}", request_id, timestamp, operation);

        let mut mac = HmacSha256::new_from_slice(shared_secret).unwrap();
        mac.update(data_to_sign.as_bytes());
        let signature = hex::encode(mac.finalize().into_bytes());

        let token = format!("{}:{}:{}:{}", request_id, timestamp, operation, signature);

        let result = validate_token(&token, shared_secret);
        assert!(result.is_err());
        assert!(matches!(result, Err(TokenError::TokenExpired(_))));
    }

    #[test]
    fn test_validate_token_invalid_signature() {
        let shared_secret = b"test-secret-key-32-bytes-long!";
        let request_id = "test-request-id";
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_secs() as i64;
        let timestamp = now;
        let operation = "test-operation";

        let token = format!("{}:{}:{}:{}", request_id, timestamp, operation, "invalid-signature");

        let result = validate_token(&token, shared_secret);
        assert!(result.is_err());
        assert!(matches!(result, Err(TokenError::InvalidSignature)));
    }
}
